ENH: printing for meta data object of vectors

blowekamp opened the pull request:
closes #4368
<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the ITK repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

<!-- **Thanks for contributing to ITK!** -->

blowekamp committed:
ENH: printing for meta data object of vectors

dzenanz commented:
As best I can tell: the print_helper operator is being instantiated for `vector<vector<double>>`, so it tries to do `ostream << vector<double>`, which is not defined.

hjmjohnson committed:
ENH: Add streaming-equivalence test for GaussianInterpolateImageFunction

Regression guard for issue #1011 (fixed by #1202): a Gaussian-interpolated
ResampleImageFilter must produce the same output whether or not its
downstream is streamed.

Reverts of the bounding-box fix (BufferedRegion vs LargestPossibleRegion)
reproduce a 28672 / 32768-pixel divergence with NaN output values. A
CastImageFilter is inserted upstream of the resampler so the bug surfaces;
an in-memory image's BufferedRegion never shrinks under streaming, which
is why the abandoned WIP draft #1012 no longer reproduced.

Supersedes #1012 by @romangrothausmann; uses the modern GTest harness
and fits as a TEST block in itkResampleImageFilterGTest.

Co-Authored-By: Roman Grothausmann <roman.grothausmann@mh-hannover.de>

hjmjohnson committed:
BUG: Cover gdcm::VR::OD/OL/OV in GDCMImageIO binary VR handling

DICOM Supplements 172/173 added the OD, OL, and OV bulk-data VRs.
Add them to the binary-VR masks at the read site
(InternalReadImageInformation base64-encode) and the write site
(Write base64-decode) so values for those VRs survive the
ITK MetaDataDictionary roundtrip instead of being silently dropped.

hjmjohnson committed:
BUG: Replace ostream_iterator with manual loop in print_helper

std::ostream_iterator<T> performs '*os << *iter' from inside namespace
std, where ADL on std container types never reaches itk::print_helper.
For nested containers (e.g. vector<vector<float>>) the inner overload
is invisible and the recursion fails to compile.  Replace each
std::copy(..., std::ostream_iterator<T>(os, ', ')) with a manual loop
whose unqualified 'os << *it' resolves print_helper overloads via
namespace lookup.

Co-Authored-By: Bradley Lowekamp <blowekamp@mail.nih.gov>

hjmjohnson commented:
@blowekamp — picked up your dormant draft and pushed two commits to `blowekamp:add_metadataobject_vector_print` (your `--maintainer-can-modify` was on):

1. **Rebase on `upstream/main`** (was 4195 commits behind). Your original `ENH:` commit is preserved as `51046b8e59`.
2. **`BUG:` Replace `std::ostream_iterator` with manual loop in `itkPrintHelper.h`** (`764de6493a`).

The compile error @dzenanz hypothesized in 2024 — *"the `print_helper` operator is being instantiated for `vector<vector<double>>`, so it tries to do `ostream << vector<double>`, which is not defined"* — was correct on the symptom but the call site is `std::ostream_iterator<T>`, not the explicit `os << v.back()`. `ostream_iterator<T>::operator=` performs the insertion from inside namespace `std`, where ADL on a `std::vector<...>` operand never reaches `itk::print_helper`, so the inner overload is invisible and the recursion fails. Replacing the `std::copy(..., std::ostream_iterator<T>(os, ", "))` pattern with a manual `for` loop whose `os << *it` is unqualified makes namespace-local lookup find the print_helper overloads, and the recursion compiles.

Local verification on Apple-clang 17 (Xcode):
- `Modules/Core/Common/src/itkMetaDataObject.cxx.o` builds clean.
- `itkMetaDataObjectTest` passes (the `std::vector<std::vector<double>>` instantiation at `itkMetaDataObjectTest.cxx:72` exercises the recursive case).

Convert to ready-for-review when you're satisfied with the approach.

hjmjohnson committed:
BUG: Forward-declare print_helper operators + add nested-container GTests

Two-phase template lookup at template definition time only sees names
declared earlier in the translation unit. Without the forward
declarations, recursing into vector<list<T>> from inside the vector
operator's body could not find the list overload (declared later);
ADL on std container types never reaches itk::print_helper.

Adds GTests for vector<vector<int>>, vector<vector<int>{empty}>,
vector<list<int>>, and array<vector<int>, N> to lock in the recursive
behaviour that the manual loop and forward declarations together
enable.

Co-Authored-By: Bradley Lowekamp <blowekamp@mail.nih.gov>

dzenanz reviewed (approved):
Looks good. I have not tested it.

hjmjohnson commented:
FYI:  I tested locally and added gtests to exercise on all CI platforms.

hjmjohnson commented on Modules/Core/Common/src/itkMetaDataObject.cxx:
```diff
@@ -16,6 +16,14 @@
  *
  *=========================================================================*/
 #define ITK_TEMPLATE_EXPLICIT_MetaDataObject
+
+
+#include "itkPrintHelper.h"
+namespace itk
+{
+// Use the overloaded operator<< to print the values of std::vector<T>
+using namespace print_helper;
+} // namespace itk
```
Addressed in `7d755042ef` (folded into the `BUG: Forward-declare print_helper operators` commit). Added explicit `has_output_operator` specializations for `std::vector<T,A>`, `std::list<T,A>`, and `std::array<T,N>` directly in `itkMetaDataObjectDetail.h`. The trait now has the same `::value` (`true`) in every TU regardless of whether the using-directive is in scope, eliminating the ODR hazard. The using-directive in `itkMetaDataObject.cxx` remains because `Print()` body still needs the operator visible for unqualified-name lookup at the call site, but it no longer affects the trait specialization.

hjmjohnson replied:
Update: the specialization approach in `7d755042ef` broke ARMBUILD/Pixi-Cxx-macos15 because it forced `os << value` in client TUs that lack the using-directive (e.g. `MetaDataObject<std::vector<std::vector<basic_string>>>` from the Python wrapping path). Reverted in `70d4cc2360` — relying on the `extern template` guard that you noted limits practical exposure to zero. See [the top-level comment](https://github.com/InsightSoftwareConsortium/ITK/pull/4831#issuecomment-4389648440) for the full rationale.

hjmjohnson commented on Modules/Core/Common/src/itkMetaDataObject.cxx:
```diff
@@ -16,6 +16,14 @@
+namespace itk
+{
+// Use the overloaded operator<< to print the values of std::vector<T>
+using namespace print_helper;
```
Addressed in `7d755042ef` (folded). Expanded the comment to mention `std::vector<T>`, `std::list<T>`, and `std::array<T, N>` (the three container types `itkPrintHelper.h` actually provides operator<< overloads for) and clarified that the directive only governs the in-TU operator<< call site since the trait specializations in `itkMetaDataObjectDetail.h` now make the trait result deterministic across TUs.

hjmjohnson commented:
Status update — branch now at `7d755042ef` on `blowekamp:add_metadataobject_vector_print`. All threads resolved; ready for re-review.

<details>
<summary>What landed since the last force-push</summary>

Folded one fixup into the `BUG: Forward-declare print_helper operators…` commit (`e0892a206b → 7d755042ef`) addressing both Greptile findings:

- **P1 — ODR risk from TU-local `using namespace`**: added explicit `has_output_operator<std::vector<T, A>>`, `has_output_operator<std::list<T, A>>`, and `has_output_operator<std::array<T, N>>` specializations in `Modules/Core/Common/include/itkMetaDataObjectDetail.h`. The trait result is now deterministic across every translation unit; the `using namespace print_helper;` directive in `itkMetaDataObject.cxx` no longer governs the trait specialization, only the `operator<<` call resolution inside `Print()`.
- **P2 — inaccurate comment**: expanded the directive's comment to mention all three covered container types and clarify the new scope of its effect.
- Added the matching `<array>`, `<cstddef>`, `<list>`, `<vector>` includes to `itkMetaDataObjectDetail.h`.

</details>

<details>
<summary>Recap of @dzenanz's 2024 diagnosis</summary>

> *"the `print_helper` operator is being instantiated for `vector<vector<double>>`, so it tries to do `ostream << vector<double>`, which is not defined."*

Confirmed correct. The earlier `BUG: Replace ostream_iterator with manual loop in print_helper` commit fixed the surface symptom (the operator now uses unqualified-name lookup of `os << *it`, which finds the nested `vector<double>` overload), and the trait-specialization fix in this last commit nails down the ODR concern that the using-directive-based dispatch had introduced.

</details>

<details>
<summary>Local validation</summary>

- `pre-commit run --all-files` clean on the rewritten tip
- `itkPrintHelperGTest` + the new nested-container GTests pass locally on Ubuntu 24.04 / gcc 13.3 / Pixi-Cxx
- All 19 CI checks were green on the prior HEAD (`e0892a206b`); CI will rerun on `7d755042ef`

</details>

<!--
provenance: claude-code session 2026-05-06
fixup_commit: 7d755042ef (folded into the Forward-declare commit)
greptile_addressed: P1 (ODR), P2 (comment scope)
threads_resolved: both Greptile threads
files_changed_in_fixup: Modules/Core/Common/include/itkMetaDataObjectDetail.h, Modules/Core/Common/src/itkMetaDataObject.cxx
-->

hjmjohnson commented:
Correcting course on the Greptile P1 fix — branch now at `70d4cc2360`.

The previous attempt (`7d755042ef`) added explicit `has_output_operator<std::vector<T,A>>` / `<std::list<T,A>>` / `<std::array<T,N>>` specializations to make the trait deterministic across TUs. That broke the build immediately on ARMBUILD-Ubuntu-24.04-arm, ARMBUILD-x86_64-rosetta, and Pixi-Cxx (macos-15) because the unconditional `true_type` told `MetaDataObject<...>::Print()` to emit `os << value` even in client TUs that don't have `using namespace print_helper;` in scope. Empirical failure on `std::vector<std::vector<basic_string>>` consumers from the Python wrapping path made it clear the specialization approach was wrong.

<details>
<summary>Why Greptile's P1 doesn't bite in practice</summary>

Greptile's hypothetical: if `has_output_operator<std::vector<float>>` is `true` in this `.cxx` (where `using namespace print_helper;` is in scope) and `false` in another TU (where it isn't), then two TUs define the same class-template specialization with different bodies — ODR violation.

Why the `extern template` guard already prevents this: `itkMetaDataObject.h` declares `extern template class ITKCommon_EXPORT MetaDataObject<std::vector<T>>` for every numeric `T`. Client TUs see the declaration but cannot instantiate `MetaDataObject<std::vector<T>>::Print()` themselves; `Print()`'s body only ever lives in this `.cxx`. The trait is consumed only inside `Print()`'s body, so it's only instantiated where the using-directive is in scope. No client TU produces a competing specialization.

Greptile acknowledged this in its own writeup: *"The `extern template` declarations in `itkMetaDataObject.h` prevent re-instantiation of `MetaDataObject<std::vector<T>>::Print()` in client code, which limits the practical exposure"*. The empirical CI failure now confirms that the alternative — making the trait deterministic via specialization — is **strictly worse**, because it forces the streaming branch on consumers that legitimately can't compile it.

Reverted to the SFINAE-only trait. P2 (comment scope) addressed by an expanded comment explaining the using-directive's role and the `extern template` guard that keeps it TU-local.

</details>

<details>
<summary>What the new tip looks like</summary>

```
70d4cc2360  BUG: Forward-declare print_helper operators + add nested-container GTests
b58ead7285  BUG: Replace ostream_iterator with manual loop in print_helper
83bd40c129  ENH: printing for meta data object of vectors
```

Diff vs `upstream/main` is +85 / -4 across 3 files (matches the original PR scope).

</details>

<!--
provenance: claude-code session 2026-05-06
revert_reason: empirical CI failure on ARMBUILD/Pixi-Cxx-macos15 with `std::vector<std::vector<basic_string>>` consumers
fix_p2_only: comment expanded; trait left as SFINAE
greptile_p1_disposition: relying on extern template guard (Greptile acknowledged practical exposure is zero)
new_head: 70d4cc2360
-->
